File: src/main/java/io/protobase/f7/formulas/logic/XOR.java
package io.protobase.f7.formulas.logic;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;
import io.protobase.f7.utils.Mappers;

import java.util.Arrays;
import java.util.Objects;
import java.util.function.BiFunction;
import java.util.function.Function;

public class XOR extends AbstractFormula {
  public static FormulaName NAME = FormulaName.XOR;
  public static XOR SELF = new XOR();

  public XOR() {
    super();
  }

  public XOR(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkAtLeastLength(values.length, 1, NAME);
    return Arrays.stream(values)
        .map(lookup)
        .flatMap(Mappers::toFlatStream)
        .filter(Objects::nonNull)
        .map(Converters::toBoolean)
        .reduce(Boolean::logicalXor)
        .orElse(false);
  }
}

File: src/main/js/formulas/logic/XOR.ts
import { isNotNull } from "../../utils/Other";
import { ValueException } from "../../errors/ValueException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { Mappers } from "../../utils/Mappers";
import { Reducers } from "../../utils/Reducers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class XOR extends AbstractFormula {
  static SELF: XOR = new XOR();
  NAME = FormulaName.OR;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkAtLeastLength(values.length, 1, this.NAME);
    const filteredValues = values
      .map(this.lookup)
      .map(Mappers.toFlatStream)
      .reduce(Reducers.join)
      .filter(isNotNull);
    if (filteredValues.length === 0) {
      return new ValueException("OR has no valid input data.");
    }
    return filteredValues.map(Converters.toBoolean).reduce(Reducers.logicalXOr, false);
  }
}
